src/Traits/ErrorTrait.php

```diff
@@ -35,67 +50,189 @@ trait ErrorTrait
      * @param string $varName The name of the userinput variable, to be used in the error message.
      * @param string|string[] $classNames The classnames that should be tested against
      * @param mixed $userInput The input that should be tested
+     *
      * @throws TypeError
      */
-    private function assertClass(string $varName, $classNames, $userInput): void
+    private static function assertClass(string $varName, $classNames, $userInput): void
     {
         if (!is_array($classNames)) {
             $classNames = [$classNames];
         }
 
+        if (!self::isClass($classNames, $userInput)) {
+            throw self::typeError($varName, $classNames, $userInput);
+        }
+    }
+
+    /**
+     * Get debug type method stolen and refactored from the symfony polyfill
+     *
+     * @see https://github.com/symfony/polyfill/blob/main/src/Php80/Php80.php
+     */
+    public static function getDebugType($value): string
+    {
+        return self::detectScalar($value)
+            ?? self::detectClass($value)
+            ?? self::detectResource($value)
+            ?? 'unknown';
+    }
+
+    /**
+     * @param string[] $classNames
+     * @param mixed $userInput
+     * @return bool
+     */
+    private static function isClass(array $classNames, $userInput): bool
+    {
         foreach ($classNames as $class) {
-            if ($userInput instanceof $class) {
-                return;
+            if (is_a($userInput, $class) || $class === self::getDebugType($userInput)) {
+                return true;
             }
         }
 
-        throw new TypeError(
-            $this->getTypeErrorText(
-                $varName,
-                $classNames,
-                $userInput
-            )
-        );
+        return false;
     }
 
     /**
-     * Give a nice error message about $userInput not being an object with one of the $classNames types.
+     * Validates the name to see if it can be used as a parameter or variable.
      *
-     * @param string $varName The name of the variable to be used in the message (without trailing '$')
-     * @param array $classNames The class names that should be mentioned in the message
-     * @param mixed $userInput The input that has been given
+     * @see https://neo4j.com/docs/cypher-manual/current/syntax/naming/#_naming_rules
+     *
+     * @param string $name
+     *
+     * @return void
+     */
+    private static function assertValidName(string $name): void
+    {
+        $name = trim($name);
+
+        if ($name === "") {
+            throw new InvalidArgumentException("A name cannot be an empty string");
+        }
+
+        if (!ctype_alnum(str_replace('_', '', $name))) {
```

Comment:
Just a nitpick, but do note that `ctype_alnum` does not accept non-ascii alphanumeric (UTF-8) characters, even though these are allowed for names by Neo4j. (I believe).

Better would be something like `if (!preg_match('/^\p{L}[\p{L}0-9_]*$/u', $name)) {throw Exception('should be alphanumeric or underscore and start with alphabetic');}` or something (This also checks for first character to be alphabetic)